src/app/store/machine/types/actions.ts

```diff
@@ -236,11 +242,55 @@ export enum FetchSortDirection {
   Descending = "descending",
 }
 
-// TODO: Define the specific filters that the API supports:
-// https://github.com/canonical-web-and-design/app-tribe/issues/1149
-export type FetchFilters = {
-  [x: string]: string | number | boolean | (string | number)[];
-};
+type Filters = {
+  free_text: string;
+  system_id: Machine["system_id"];
+  hostname: Machine["hostname"];
+  description: Machine["description"];
+  distro_series: Machine["distro_series"];
+  osystem: Machine["osystem"];
+  error_description: Machine["error_description"];
+  mac_address: NetworkInterface["mac_address"];
+  domain: Domain["name"];
+  agent_name: string;
+  fabric_classes: Fabric["class_type"];
+  fabrics: Machine["fabrics"];
+  zone: Machine["zone"]["name"];
+  pool: ResourcePool["name"];
+  arch: string;
+  tags: Tag["name"];
+  vlans: NodeVlan["name"];
+  pod: ModelRef["name"];
+  pod_type: string;
+  owner: Machine["owner"];
+  subnets: Subnet["name"];
+  ip_addresses: NodeIpAddress["ip"];
+  spaces: Space["name"];
+  workloads: string;
+  status: FetchNodeStatus;
+};
+
+type ExcludeFilters = {
+  not_fabric_classes: Filters["fabric_classes"];
+  not_fabrics: Filters["fabrics"];
+  not_in_zone: Filters["zone"];
+  not_in_pool: Filters["pool"];
+  not_arch: Filters["arch"];
+  not_tags: Filters["tags"];
+  not_system_id: Filters["system_id"];
+  not_pod: Filters["pod"];
+  not_pod_type: Filters["pod_type"];
+  not_owner: Filters["owner"];
+  not_subnets: Filters["subnets"];
+  not_vlans: Filters["vlans"];
+  not_ip_addresses: Filters["ip_addresses"];
+  not_distro_series: Filters["distro_series"];
+  not_osystem: Filters["osystem"];
+};
+
+type ArrayOrValue<T> = { [P in keyof T]?: T[P] | Array<T[P]> };
+
+export type FetchFilters = Partial<ArrayOrValue<Filters & ExcludeFilters>>;
```

Comment:
It looks like not everything can be an array:

```python
NODE_FILTERS = {
    "fabric_classes": (
        "current_config__interface__vlan__fabric__class_type",
        _match_any,
    ),
    "fabrics": (
        "current_config__interface__vlan__fabric__name",
        _match_any,
    ),
    "vlans": ("current_config__interface__vlan", _match_any),
    "zone": ("zone__name", _match_any),
    "pool": ("pool__name", _match_any),
    "mem": ("memory", _match_gte),
    "cpu_count": ("cpu_count", _match_gte),
    "cpu_speed": ("cpu_speed", _match_gte),
    "arch": ("architecture", _match_any),
    "link_speed": ("current_config__interface__link_speed", _match_gte),
    "tags": ("tags__name", _match_all),
    "subnets": (
        "current_config__interface__ip_addresses__subnet",
        _match_all,
    ),
    "pod": ("bmc__name", _match_any),
    "pod_type": ("bmc__power_type", _match_any),
    "owner": ("owner__username", _match_any),
    "power_state": ("power_state", _match_any),
}

NODE_EXCLUDES = {
    "not_fabric_classes": (
        "current_config__interface__vlan__fabric__class_type",
        _match_any,
    ),
    "not_fabrics": (
        "current_config__interface__vlan__fabric__name",
        _match_any,
    ),
    "not_vlans": ("current_config__interface__vlan", _match_any),
    "not_in_zone": ("zone__name", _match_any),
    "not_in_pool": ("pool__name", _match_any),
    "not_arch": ("architecture", _match_any),
    "not_tags": ("tags__name", _match_any),
    "not_subnets": (
        "current_config__interface__ip_addresses__subnet",
        _match_any,
    ),
    "not_system_id": ("system_id", _match_any),
    "not_pod": ("bmc__name", _match_any),
    "not_pod_type": ("bmc__power_type", _match_any),
    "not_owner": ("owner__username", _match_any),
    "not_power_state": ("power_state", _match_any),
}
```

Looking at those lists we might need to define the types for filters with `_match_any` and `_match_all` to be arrays but `_match_gte` to be a single value.